### ISSUE TITLE
`prefer_void_to_null` should not warn when a class implements a contract's nullable type as `Null`

### ISSUE THREAD

**Issue by jellynoone** [dart-lang/linter#3076]
**Describe the issue**
Currently when a class implements an interface that defines a nullable type as a return type or property type and makes them explicitly `Null`, a `prefer_void_to_null` warning is triggered.

**To reproduce**
```dart
abstract class Contract {
  String? get prop;
  bool? doSomething();
}

class ContractSuccessImplementation implements Contract {
  @override
  final String prop;

  ContractSuccessImplementation(this.prop);

  @override
  bool doSomething() => prop.isEmpty;
}

class ContractErrorImplementation implements Contract {
  @override
  Null get prop => null;

  @override
  Null doSomething() => null;
}

class ContractErrorImplementation2 implements Contract {
  @override
  final Null prop = null;

  @override
  Null doSomething() => null;
}
```

**Expected behavior**
Since `Null` is part of the contract reducing the type to just `Null` in an implementation should not raise a lint, since it cannot be replaced with `void`.

Sometimes you know for an implementation that the field / return type is going to be null. It would be great to be able to type that without lint warnings.

**Additional**
`Dart SDK version: 2.14.4 (stable)`

**Comment by eernstg** [comment]
+1,

A covariant specialization of the return type of a method from `T?` to `Null` precisely indicates that the subtype implementation will return null (that is, it will "use the `?` part of `T?`, not the `T` part"), and this will help developers understand the actual behavior of the invocation at call sites where the receiver is known to have the more specialized type (here: `ContractErrorImplementation` or `ContractErrorImplementation2`): They won't waste the resources (readability, code size, run-time performance) on testing whether the returned object `is T` and acting on it, because we know it isn't.

You could argue that this is nevertheless an error prone typing, because a value of type `Null` could be used in many different contexts without error (I believe that was the main reason why the use of `Null` rather than `void` was flagged as not recommended), but that reason is much less convincing today: With null safety, the number of situations where a value of type `Null` can be passed is simply so much smaller.

(An aside: What we'd ideally have is the ability to mark arbitrary types as "voidy", so `void<Null>` would be a type which is treated like `void` in that any attempt to use the value would be flagged as an error (there's a whitelist of exceptions, but it's typically an error), and it would be equivalent to `Null` according to the subtyping relationship and the soundness properties (so we'd ensure that the actual value of type `void<Null>` is `null`, just like anything with type `Null`). We don't have that, but it illustrates that the ability to flag usages of a particular value could be provided for all types, and we would then be able to prevent unwanted usages of an object also in this case.)

**Comment by srawlins** [comment]
I believe this was implemented in https://github.com/dart-lang/linter/commit/770428acc1a77c88ddc0d93729abb67eae7756ec, which landed a few months before this issue was opened.  But everything looks good. Please open if this is still an issue.
